free_fleet_adapter/free_fleet_adapter/nav2_robot_adapter.py

```diff
@@ -0,0 +1,290 @@
+import rclpy
+import rmf_adapter.easy_full_control as rmf_easy
+from tf_transformations import quaternion_from_euler
+import zenoh
+
+
+class Nav2RobotAdapter:
+    def __init__(
+        self,
+        name: str,
+        configuration,
+        robot_config_yaml,
+        node,
+        zenoh_session,
+        fleet_handle,
+        tf_buffer
+    ):
+        self.name = name
+        self.execution = None
+        self.update_handle = None
+        self.configuration = configuration
+        self.robot_config_yaml = robot_config_yaml
+        self.node = node
+        self.zenoh_session = zenoh_session
+        self.fleet_handle = fleet_handle
+        self.tf_buffer = tf_buffer
+
+        self.nav_goal_id = None
+        self.map = self.robot_config_yaml['initial_map']
+
+        # TODO(ac): Only use full battery if sim is indicated
+        self.battery_soc = 1.0
+
+        def _tf_callback(sample: zenoh.Sample):
+            try:
+                transform = TFMessage.deserialize(sample.payload)
+            except Exception as e:
+                self.node.get_logger().debug(
+                    f'Failed to deserialize TF payload: {type(e)}: {e}'
+                )
+                return None
+            for zt in transform.transforms:
+                time = rclpy.time.Time(
+                    seconds=zt.header.stamp.sec,
+                    nanoseconds=zt.header.stamp.nanosec
+                )
+                t = TransformStamped()
+                t.header.stamp = time.to_msg()
+                t.header.stamp
+                t.header.frame_id = namespacify(zt.header.frame_id,
+                                                self.name)
+                t.child_frame_id = namespacify(zt.child_frame_id,
+                                               self.name)
+                t.transform.translation.x = zt.transform.translation.x
+                t.transform.translation.y = zt.transform.translation.y
+                t.transform.translation.z = zt.transform.translation.z
+                t.transform.rotation.x = zt.transform.rotation.x
+                t.transform.rotation.y = zt.transform.rotation.y
+                t.transform.rotation.z = zt.transform.rotation.z
+                t.transform.rotation.w = zt.transform.rotation.w
+                self.tf_buffer.set_transform(t, f'{self.name}_RobotAdapter')
+
+        self.tf_sub = self.zenoh_session.declare_subscriber(
+            namespacify('tf', self.name),
+            _tf_callback
+        )
+
+        def _battery_state_callback(sample: zenoh.Sample):
+            battery_state = SensorMsgs_BatteryState.deserialize(sample.payload)
+            self.battery_soc = battery_state.percentage
+
+        self.battery_state_sub = self.zenoh_session.declare_subscriber(
+            namespacify('battery_state', name),
+            _battery_state_callback
+        )
+
+    def _make_random_goal_id(self):
+        return np.random.randint(0, 255, size=(16)).astype('uint8').tolist()
```

Comment:
Nit, though its random and a size 16 list, I wonder if there's a more reliable way to prevent overlapping goal IDs.

Reply:
I think this should be random enough, it'll be 256^16 combinations, but I'll check out how `nav2` usually does it and use a similar method perhaps